MarvelApiClient/src/test/java/com/karumi/MarvelApiClient/CharacterApiClientTest.java

```diff
@@ -0,0 +1,251 @@
+    CharacterApiClient characterApiClient = givenCharacterApiClient();
+    enqueueMockResponse("getCharacters.json");
+
+    MarvelResponse<CharactersDto> characters = characterApiClient.getAll(0, ANY_LIMIT);
+
+    assertBasicMarvelResponse(characters);
+    CharactersDto charactersDto = characters.getResponse();
+    assertEquals(10, charactersDto.getCount());
+    assertEquals(10, charactersDto.getLimit());
+    assertEquals(0, charactersDto.getOffset());
+    assertEquals(1485, charactersDto.getTotal());
+
+    CharacterDto firstCharacter = charactersDto.getCharacters().get(0);
+    assertIs3dManCharacter(firstCharacter);
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void shouldReturnAnIllegalExceptionWhenCallGetCharacterWithoutId() throws Exception {
+    CharacterApiClient characterApiClient = givenCharacterApiClient();
+
+    characterApiClient.getCharacter(INVALID_CHARACTER_ID);
+  }
+
+  @Test public void shouldReturnAValidUrlWhenGetCharacterHasAValidId() throws Exception {
+    CharacterApiClient characterApiClient = givenCharacterApiClient();
+    enqueueMockResponse("getCharacter.json");
+
+    characterApiClient.getCharacter(THREED_MAN_CHARACTER_ID);
+
+    assertRequestSentToContains("characters/" + THREED_MAN_CHARACTER_ID);
+  }
+
+  @Test(expected = MarvelApiException.class)
+  public void shouldReturnMarvelExceptionWhenTheIdDoesNotExist() throws Exception {
+    CharacterApiClient characterApiClient = givenCharacterApiClient();
+    enqueueMockResponse(404, "{\"code\":404,\"status\":\"We couldn't find that character\"}");
+
+    try {
+      characterApiClient.getCharacter(ANY_NOT_FOUND_ID);
+    } catch (MarvelApiException e) {
+      assertEquals("404", e.getMarvelCode());
+      assertEquals("We couldn't find that character", e.getMessage());
+      throw e;
+    }
+  }
+
+  @Test public void shouldReturnAvalidResponseWhenCallGetCharacterWithValidId() throws Exception {
+    CharacterApiClient characterApiClient = givenCharacterApiClient();
+    enqueueMockResponse("getCharacter.json");
+
+    MarvelResponse<CharacterDto> character = characterApiClient.getCharacter(
+        THREED_MAN_CHARACTER_ID);
+
+    assertBasicMarvelResponse(character);
+
+    CharacterDto response = character.getResponse();
+    assertIs3dManCharacter(response);
+  }
+
+  private List<Integer> getAnyEvents() {
+    List<Integer> events = new ArrayList<>();
+    events.add(ANY_EVENT_1);
+    events.add(ANY_EVENT_2);
+    return events;
+  }
+
+  private CharacterApiClient givenCharacterApiClient() {
+
+    Retrofit retrofit = new Retrofit.Builder().baseUrl(getBaseEndpoint())
+        .addConverterFactory(GsonConverterFactory.create())
+        .build();
+
+    MarvelApiClient marvelApiClient =
+        new MarvelApiClient.Builder(ANY_PUBLIC_KEY, ANY_PRIVATE_KEY).baseUrl(ANY_URL)
+            .retrofit(retrofit)
+            .build();
+    return new CharacterApiClient(marvelApiClient);
```

Comment:
If this is the library usage we should review it. I'd like to change it to be able to use it doing something like this:

```java
MarvelApiClient.setAuthConfig(ANY_PUBLIC_KEY, ANY_PRIVATE_KEY)
MarvelApiClient.setBaseUrl(ANY_URL)
MarvelApiClient.setRetrofit(retrofit)
CharacterApiClient characterApiClient =MarvelApiClient.getCharactersAPiClient()
```

If I'm not wrong, doing this you can remove the singleton you have inside :)